70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+			break;
+		}
+		
+		fillGridData();
+	}
+
+	void fillGridData()
+	{
+		m_gridData.particleInitSize = m_gridData.particleInitMax - m_gridData.particleInitMin;
+		float32_t4 simAreaSize = m_gridData.gridSize;
+		simAreaSize *= m_gridData.gridCellSize;
+		m_gridData.worldMin = float32_t4(0.f);
+		m_gridData.worldMax = simAreaSize;
+		numGridCells = m_gridData.gridSize.x * m_gridData.gridSize.y * m_gridData.gridSize.z;
+		numParticles = m_gridData.particleInitSize.x * m_gridData.particleInitSize.y * m_gridData.particleInitSize.z * particlesPerCell;
+	}
+
+	smart_refctd_ptr<IGPUShader> compileShader(const std::string& filePath, const std::string& entryPoint = "main")
+	{
+		IAssetLoader::SAssetLoadParams lparams = {};
+		lparams.logger = m_logger.get();
+		lparams.workingDirectory = "";
+		auto bundle = m_assetMgr->getAsset(filePath, lparams);
+		if (bundle.getContents().empty() || bundle.getAssetType() != IAsset::ET_SHADER)
+		{
+			m_logger->log("Shader %s not found!", ILogger::ELL_ERROR, filePath);
+			exit(-1);
+		}
+		
+		const auto assets = bundle.getContents();
+		assert(assets.size() == 1);
+		smart_refctd_ptr<ICPUShader> shaderSrc = IAsset::castDown<ICPUShader>(assets[0]);
+
+		smart_refctd_ptr<ICPUShader> shader = shaderSrc;
+		if (entryPoint != "main")
+		{
+			auto compiler = make_smart_refctd_ptr<asset::CHLSLCompiler>(smart_refctd_ptr(m_system));
+			CHLSLCompiler::SOptions options = {};
+			options.stage = shaderSrc->getStage();
+			if (!(options.stage == IShader::E_SHADER_STAGE::ESS_COMPUTE || options.stage == IShader::E_SHADER_STAGE::ESS_FRAGMENT))
+				options.stage = IShader::E_SHADER_STAGE::ESS_VERTEX;
+			options.targetSpirvVersion = m_device->getPhysicalDevice()->getLimits().spirvVersion;
+			options.spirvOptimizer = nullptr;
+			options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_SOURCE_BIT;
+			options.preprocessorOptions.sourceIdentifier = shaderSrc->getFilepathHint();
+			options.preprocessorOptions.logger = m_logger.get();
+			options.preprocessorOptions.includeFinder = compiler->getDefaultIncludeFinder();
+
+			std::string dxcOptionStr[] = {"-E " + entryPoint};
+			options.dxcOptions = std::span(dxcOptionStr);
+
+			shader = compiler->compileToSPIRV((const char*)shaderSrc->getContent()->getPointer(), options);
+		}
+
+		return m_device->createShader(shader.get());
+	}
+
+	bool createBuffer(smart_refctd_ptr<IGPUBuffer>& buffer, video::IGPUBuffer::SCreationParams& params)
+	{
+		buffer = m_device->createBuffer(std::move(params));
+		if (!buffer)
+			return logFail("Failed to create GPU buffer of size %d!\n", params.size);
+
+		video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = buffer->getMemoryReqs();
+		reqs.memoryTypeBits &= m_physicalDevice->getDeviceLocalMemoryTypeBits();
+
+		auto bufMem = m_device->allocate(reqs, buffer.get());
+		if (!bufMem.isValid())
+			return logFail("Failed to allocate device memory compatible with gpu buffer!\n");
+
+		return true;
+	}
```

Comment:
there's a method in `IUtility` or `ILogicalDevice` for this IIRC (DeviceLocal in the name or something)